### PR TITLE
build: accelerate CI build by prebuild modules

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

Every CI build will spend several minutes `autoconf`, `configure`, and `make` in submodules. These are dependencies and stable across various pull requests and testing jobs. Avoiding rebuild them can significantly save time in CI jobs, especially those simple ones such as warnings tests. For review tests, it can shorten the developer feedback time. For nightly tests, consider that we run hundreds of CI tests, they save quite some energy as well.

* add `maint/prebuild_modules.sh` that can be used to pre-build modules, create `modules.tar.gz`
* If `moldules.tar.gz` exist and run `autogen.sh -quick`, both autogen and configure will skip pre-built modules

## Why not use external libs
That's one way to avoid rebuilding modules. Potential points --
* Maintaining external install paths can be a hassle.
* We'd like for our CI tests to test the embedded module versions. It is difficult to ensure the external installations are always up-to-date.
* Other than avoiding rebuilding, the rest of the build is *exactly* like an embedded build. Linking externally may have different behavior from an embedded build.
* We have separate "external" tests already.

## Further ideas:
* split configure of `test/mpi` -- https://github.com/pmodels/mpich/pull/5067
   Not all jobs need to run the full test suite.
* create separate `Fortran and `io` tests and skip both for most review jobs
* Add options to do "amalgamation" -- [ref](https://www.sqlite.org/amalgamation.html)
   It will cut significant make time considering that we have so many repeated inline headers

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/custom  --disable-fortran --disable-romio
do-quick

**Comment by hzhou** [comment]
test:mpich/custom
do-quick

**Comment by hzhou** [comment]
test:mpich/custom
do-quick

**Comment by hzhou** [comment]
test:mpich/custom
do-quick

**Comment by hzhou** [comment]
test:mpich/custom
do-quick

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
do-quick

**Comment by hzhou** [comment]
test:mpich/custom --with-datatype-engine=dataloop
netmod: ch4:ofi
do-quick

**Comment by hzhou** [comment]
test:mpich/custom --with-datatype-engine=dataloop
netmod: ch4:ofi
do-quick

**Comment by hzhou** [comment]
test:mpich/custom --with-datatype-engine=dataloop
netmod: ch4:ofi
do-quick

**Comment by wesbland** [comment]
> Apparent the non-performance-critical parts in ch4 is really slow compared to ch3

That is a shockingly high number.

**Comment by wesbland** [comment]
How does the modules tarball get updated if the submodule gets updated?

**Comment by hzhou** [comment]
> 
> 
> How does the modules tarball get updated if the submodule gets updated?

The modules tarball is updated with Jenkins nightly or cron job. Hopefully, the modules are fairly stable so a bit out-of-date usually is not an issue. We will have test parameters to do the regular build (skipping the pre-built modules) if necessary.

**Comment by wesbland** [comment]
Having a way to tell those tarballs to update from a PR would probably be nice for days where we merge a submodule update and want to use it at the same time would be nice. Up to you though.

**Comment by raffenet** [comment]
> test:mpich/custom netmod: ch4:ofi do-quick
> 
> autogen + configure	3:18
> make	2:40
> make install	0:13
> test suite (2391 tests)	2:11:27
> Apparent the non-performance-critical parts in ch4 is really slow compared to ch3

So with ofi:sockets, we disable auto progress in libfabric even though it is enabled by default. My recollection is that test suites runs are much faster with it on...

**Comment by hzhou** [comment]
test:mpich/custom --with-datatype-engine=dataloop
netmod: ch4:ofi
do-quick

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

File: confdb/aclocal_modules.m4 [hunk 1]
```diff
@@ -77,9 +77,11 @@ AC_DEFUN([PAC_CONFIG_HWLOC],[
     if test "$with_hwloc" = "embedded" ; then
         m4_if(hwloc_embedded_dir, [modules/hwloc], [
             dnl ---- the main MPICH configure ----
-            PAC_CONFIG_HWLOC_EMBEDDED([$VISIBILITY_CFLAGS])
-            hwlocsrcdir="${main_top_builddir}/modules/hwloc"
             hwloclib="${main_top_builddir}/modules/hwloc/hwloc/libhwloc_embedded.la"
+            if test ! -e "$hwloclib" ; then
+                hwlocsrcdir="${main_top_builddir}/modules/hwloc"
+                PAC_CONFIG_HWLOC_EMBEDDED([$VISIBILITY_CFLAGS])
+            fi
```
Comment (raffenet) [inline confdb/aclocal_modules.m4:84]: So when not using prebuilt modules, would this mean the modules would not get reconfigured after they are built once? That gives me some concern users could run into this and not know what is going on.

File: autogen.sh [hunk 2]
```diff
@@ -67,6 +67,138 @@ sync_external () {
     cp -pPR "$srcdir" "$destdir"
 }
 
+########################################################################
+## Step functions
+########################################################################
```
Comment (raffenet) [inline autogen.sh:72]: All this is orthogonal to the goal of this PR. I'd rather consider it separately.

File: autogen.sh [hunk 3]
```diff
@@ -309,67 +309,53 @@ for arg in "$@" ; do
...
-
         -verbose-autoreconf|--verbose-autoreconf)
             autoreconf_args="-vif"
             export autoreconf_args
             ;;
 
-	-with-errmsgs|--with-errmsgs)
-	    do_geterrmsgs=yes
+	-with-autotools=*|--with-autotools=*)
+	    autotoolsdir=`echo "A$arg" | sed -e 's/.*=//'`
```
Comment (raffenet) [inline autogen.sh:318]: This patch is unrelated to the goal of this PR.

**Review by raffenet — COMMENTED** [review]
My biggest concern is the change to the configuration process. The last 2 patches are just superfluous refactoring that I'd rather see separated from the functional changes.

File: confdb/aclocal_modules.m4 (hunk 1 above)
Comment (hzhou) [reply confdb/aclocal_modules.m4:84]: Yes.

File: maint/prebuild_modules.sh [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+#
+# This is to accelerate repeated CI testing.
+#
+
+git submodule update --init
+
+make_it_lean () {
+    rm -rf .git
+    find . -name '*.o' | xargs rm -f
+}
```
[the remark this answers is not in the document]
Comment (hzhou) [reply maint/prebuild_modules.sh:18]: Yes. The `.git` folder can be quite big and it is a waste of bandwidth.

File: autogen.sh (hunk 2 above)
Comment (hzhou) [reply autogen.sh:72]: Sure. I only realized later that the original `--do=xxx` is to run a particular step rather than to turn on/off a particular option. At that point, I need to switch to `--with-xxx` as options to switch on/off features. That's why many refactoring mixed in. I'll see if I can split them. Do you prefer the refactoring of `autogen.sh` goes before this PR or after? This PR needs finer options to skip module initializations, and it will be significantly cleaner if it's after the refactoring.

File: autogen.sh [hunk 5]
```diff
@@ -124,6 +124,12 @@ for arg in "$@" ; do
         do_hydra=yes
         do_hydra2=no
         do_romio=no
+
+        if test -e 'modules.tar.gz' -a ! -e 'modules/hwloc/configure.ac' ; then
```
[the remark this answers is not in the document]
Comment (hzhou) [reply autogen.sh:265]: Yes, exactly.

File: autogen.sh [hunk 6]
```diff
@@ -179,8 +179,8 @@ for arg in "$@" ; do
             esac
             ;;
 
-	-do=*|--do=*)
```
[the remark this answers is not in the document]
Comment (hzhou) [reply autogen.sh:174]: Sure.

File: autogen.sh (hunk 3 above)
Comment (hzhou) [reply autogen.sh:318]: There are `do_xxx` that runs a particular step (skipping all other steps), and there are `with_xxx` to turn on/off options. The goal of this PR is to optimistically skip steps and thus need add many `with{out}_xxx` options to fine-tune. The old code kinda mixes `with` and `do` for both, thus I find it is important to refactor and keep the semantics clear. How about I split and do the refactor first?

File: confdb/aclocal_modules.m4 (hunk 1 above)
Comment (raffenet) [reply confdb/aclocal_modules.m4:84]: This is my biggest concern. To me, it violates the principle of least astonishment if a user reconfigures the build and it silently skips the embedded modules, even if explicitly used. Is there some way we can special case the prebuilt module so that it doesn't interfere with a regular build?

Comment (hzhou) [reply confdb/aclocal_modules.m4:84]: It *does not* skip the embedded modules, it just skips the `configure` and use the already built `libhwloc-embedded.la`. It does change the old behavior that when user want to *reconfigure* libhwloc, they need `make clean` or remove the `libhwloc-embedded.la` first.

If that slight change is undesiable, I guess we can embed flag file in the prebuilt tarball so this skipping of configure only happens when we use prebuilt tarballs.

Comment (raffenet) [reply confdb/aclocal_modules.m4:84]: > If that slight change is undesiable, I guess we can embed flag file in the prebuilt tarball so this skipping of configure only happens when we use prebuilt tarballs.

That sounds like a good alternative.

File: autogen.sh (hunk 2 above)
Comment (raffenet) [reply autogen.sh:72]: Sure, let's put the refactoring first.

**Comment by hzhou** [comment]
Another insight on the testing time. A recent ch3 failure shows the timing disparity on different platforms:
```
jenkins_configure=default,label=osx
summary_junit_xml. - ./rma/win_info 4 | 87 ms | 1

jenkins_configure=default,label=freebsd64
summary_junit_xml. - ./rma/win_info 4 | 52 ms | 1

jenkins_configure=default,label=centos64
summary_junit_xml. - ./rma/win_info 4 | 0.24 s | 1

jenkins_configure=default,label=centos32
summary_junit_xml. - ./rma/win_info 4 | 0.28 s | 1

jenkins_configure=default,label=solaris
summary_junit_xml. - ./rma/win_info 4 | 1.2 s | 1
```

I suspect the difference lies in the latency probing the nic interface such as getting `hostname`.

**Comment by hzhou** [comment]
test:mpich/ch3/most
test:mpich/ch4/most

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
do-quick

File: confdb/aclocal_modules.m4 (hunk 1 above)
Comment (raffenet) [reply confdb/aclocal_modules.m4:91]: It seems like this feedback was ignored. This is still a NACK from me.

Comment (hzhou) [reply confdb/aclocal_modules.m4:91]: Ahh, sorry about that. Let me fix it.

Comment (raffenet) [reply confdb/aclocal_modules.m4:91]: What I am concerned about is a scenario where a user configures/builds MPICH, then later decides to configure/build again with different options:
```
./configure --prefix=foo && make install
```
```
./configure --prefix=bar --with-cuda && make install
```
The second configuration will use the already built submodules without CUDA support.

File: autogen.sh (hunk 2 above)
Comment (hzhou) [reply autogen.sh]: Done. Search for `PREBUILT` to find the changes.

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
do-quick

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ofi
do-quick

**Review by raffenet — APPROVED** [review]

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ucx
do-quick

**Comment by hzhou** [comment]
test:mpich/custom
netmod: ch4:ucx
do-quick
